### ISSUE TITLE
Preventing items from overlapping at a building's output doesn't work perfectly

### ISSUE THREAD

**Issue by EmeraldBlock** [tobspr/shapez.io#1016]
Whenever a belt stalls, the next item to eject from a build is meant to be positioned right behind the item in front of it (01733c4). However, an overlap can still be seen, which gets larger the closer the front item is to the building.

![image](https://user-images.githubusercontent.com/69981203/101393416-b51bb180-388c-11eb-9c47-0bb9ae74d60b.png)

**Comment by edorfaus** [comment]
I did some debugging/research via Gitpod, and it turns out that this is not just for stalled belts - the overlap happens for otherwise-empty belts as well. I set the debug options `doNotRenderStatics`, `enableEntityInspector` and `manualTickOnly`, and single-stepped forward from a fresh belt and newly created ItemProducer and got this:

![image](https://user-images.githubusercontent.com/112590/101460529-003fce00-393a-11eb-95fc-b567434c2fad.png)

(The graphics didn't load right, which often happens, but that shouldn't matter for this. The left tile is the (invisible) ItemProducer, while the right tile is a belt going away from it, which only has that one belt tile.)

This screenshot was taken on the first frame that the second shape was visible, at which point the ejector's `progress` value was 0.084, and the belt's `spacingToFirstItem` value was 0.063.

My notes show that the progress value starts at 0.042 on the first frame that the ejector has an item in it, and then increases by 0.042 for each frame (well, while doing single-stepping like this anyway), until it goes from 0.966 to 1 and hands off the item to the belt, which then has a spacing value of 0.021 (down from 1). Visually, the shape jumped a little farther than it had been doing, maybe an extra pixel or so? On the next frame, progress is 0.042 again (which is <0.05 so shape not visible), and spacing is 0.042. Next frame after that, is when I got the above screenshot.

Ten frames later, progress was 0.504 and spacing 0.273, and I got this screenshot which shows them still overlapping, but less so than before:

![image](https://user-images.githubusercontent.com/112590/101463615-f1f3b100-393d-11eb-8c39-9a3d07441e03.png)

Another 11 frames later, progress is 0.966 again (next frame should be 1), while spacing is 0.504, and the shapes are only barely overlapping:

![image](https://user-images.githubusercontent.com/112590/101465037-bb1e9a80-393f-11eb-9af2-02e44e97b56d.png)

Next frame, progress is 1, spacing is 0.525, and the shapes moved over a pixel or so but didn't (visibly) change their spacing. Notable is that the second shape is not as far out of the producer as the first one was at this point - apparently the progress limiting is working to some extent.

It's also clear by this point that the ejection progresses faster than belts move shapes. (I'm not really surprised, having previously noticed an additional factor of 2 in the ejector code.) So in effect, when trying to output a full belt, the ejector always gets backed up and has to wait for the belt to catch up. (This doesn't seem to be a problem though, except visually.)

4 frames later, the belt is about to accept the next shape - progress is still 1, spacing is 0.609:

![image](https://user-images.githubusercontent.com/112590/101466898-f326dd00-3941-11eb-97de-64bf3a553861.png)

and on the next frame, progress is still 1 but the item has moved to the belt, and the spacing is about 2.22e-16 (almost 0):

![image](https://user-images.githubusercontent.com/112590/101467072-279a9900-3942-11eb-9043-aed1e97b1899.png)

At this point, the overlap is gone, which is not surprising since both shapes are now handled by the belt. (Also, this time the shapes didn't seem to jump any extra distance.)

One thing I noticed during all this, is that the belt's spacing value is apparently in units of the size of a tile, while the ejector's progress value is (for drawing purposes) in units of the size of a shape. (At least I think so.)

So my best working theory at the moment, is that the formula that converts spacing to progress isn't quite right, and thus is not providing the correct limiting value, but I haven't actually looked into it in detail (so could be wrong). It would seem to match the behavior of having gradually less overlap though. (And might also explain that jump for the first shape, if the extra velocity being handed off is similarly not quite right.)

**Comment by edorfaus** [comment]
Hm, no, that theory was wrong. As far as I can tell, it perfectly evenly converts the range of the spacing value that we're interested in to the progress range.

However, I noticed something else that I think is the actual cause (though I'm not sure how best to fix it).

The progress value is being used to decide where to draw the shape, between the center of the tile and the edge of the tile. (Line 355/356, the 0.5 factor for tileX/Y.) This means that across the range of the progress value, it moves half a tile. (Not the size of a shape as I mentioned earlier, that was also wrong.) However, across the same range of values (after conversion from spacing), the belt moves an item 0.63 times the size of a tile (the value of `itemSpacingOnBelts`), which is further. Meaning it moves faster.

This problem is probably easier to visualize if we consider the reverse situation (as if time went backwards), where the shapes start right next to each other, and then move the same direction, but the front one moves slower than the rear one. As time passes, and they move along, they will overlap more and more.

Now, the obvious solution would be to change the code so that the ejector also moves the shape 0.63 * tileSize, but that would mean that the shape starts past the halfway point of the tile, which could make it visible out the other end of the building, moving inwards. That wouldn't be good, so we'd have to handle that somehow.

It might also be a bit tricky to set up right to keep the centering on the other axis, without a bunch of conditionals.

Maybe there's another solution that's easier, that I just haven't thought of?

**Comment by tobspr** [comment]
Comment from my fix in aa2b64e:

```
If you imagine the track between the center of the building and the center of the first belt as
a range from 0 to 1:

        Building              Belt
|         X         |         X         |
|         0...................1         |

And for example the first item on belt has a distance of 0.4 to the beginning of the belt:

        Building              Belt
|         X         |         X         |
|         0...................1         |
                            ^ item

Then the space towards this first item is always 0.5 (the distance from the center of the building to the beginning of the belt)
PLUS the spacing to the item, so in this case 0.5 + 0.4 = 0.9:

Building              Belt
|         X         |         X         |
|         0...................1         |
                            ^ item @ 0.9

Since items must not get clashed, we need to substract some spacing (lets assume it is 0.6, exact value see globalConfig.itemSpacingOnBelts),
So we do 0.9 - globalConfig.itemSpacingOnBelts = 0.3

Building              Belt
|         X         |         X         |
|         0...................1         |
                ^           ^ item @ 0.9
                ^ max progress = 0.3

Because now our range actually only goes to the end of the building, and not towards the center of the building, we need to multiply
all values by 2:

Building              Belt
|         X         |         X         |
|         0.........1.........2         |
                ^           ^ item @ 1.8
                ^ max progress = 0.6

And that's it! If you summarize the calculations from above into a formula, you get the one below.
```
